File: sim-seg.c
// explicitly simulate heteroplasmy distributions with divisions and gene conversion

#include <stdio.h>
#include <math.h>
#include <stdlib.h>

#define RND drand48()

#define NSIM 1000    // for sampling
#define NDIV 300     // maximum number of cell divisions
#define NPOP 50      // effective population size
#define MAXGC 150    // maximum gene conversion rate (AU)

int main(void)
{
  FILE *fp;
  int NGC;
  int i;
  int div, gc, sim;
  int r1, r2;
  int P[NPOP], NP[NPOP/2];
  int mcount;
  int realgc;
  int nnp;

  // open file for output
  fp = fopen("sim-seg.csv", "w");
  fprintf(fp, "ngc,div,sim,realgc,h\n");

  // loop through model parameters
  for(NGC = 0; NGC <= MAXGC; NGC += MAXGC/2)
    {
      for(sim = 0; sim < NSIM; sim++)
	{
	  realgc = 0; // tracker for actual number of gene conversion events
	  // initialise population
	  for(i = 0; i < NPOP; i++)
	    {
	      P[i] = (i < NPOP/2 ? 0 : 1);
	    }
	  // loop through divisions
	  for(div = 1; div < NDIV; div++)
	    {
	      // attempt NGC gene conversion events
	      for(gc = 0; gc < NGC; gc++)
		{
		  r1 = RND*NPOP;
		  r2 = RND*NPOP;
		  if(P[r1] != P[r2]) realgc++;
		  P[r2] = P[r1];
		}
	      nnp = 0;
	      // quick binomial division using NP as a new buffer population
	      for(i = 0; i < NPOP; i++)
		{
		  if(RND < 0.5)
		    {
		      NP[nnp] = P[i];
		      nnp++;
		    }			
		}
	      // Polya urn reamplification from NP back up to P
	      mcount = 0;
	      for(i = 0; i < NPOP; i++)
		{
		  r1 = RND*nnp;
		  P[i] = NP[r1];
		  mcount += (P[i] == 1);
		}
	      // output statistics
	      fprintf(fp, "%i,%i,%i,%i,%f\n", NGC, div, sim, realgc, (double)mcount/NPOP);
	    }
	}
    }
  fclose(fp);
  
  return 0;
}
  
	

File: sim-seg.R
### produce supplementary plots relating to influence of gene conversion on heteroplasmy distributions

library(ggplot2)
library(gridExtra)
library(see)

# read output from simulation
df = read.csv("sim-seg.csv")

# oDNA population size in model
npop = 50

# subset out content, normalising gene conversion counts by oDNA population size
dfr = data.frame()
for(ngc in unique(df$ngc)) {
  for(div in unique(df$div)) {
    sub = df[df$ngc == ngc & df$div == div,]
    dfr = rbind(dfr, data.frame(ngc = ngc, div=div, vh = var(sub$h), meangc = mean(sub$realgc)/npop/div))
  }
}

# first plot -- heteroplasmy variance with divisions and gene conversion rates
ga1 = ggplot(dfr, aes(x = div, y = vh, col = factor(ngc))) + geom_line() + xlab("Divisions") + ylab("V(h)") + labs(col = "GC rate\n(AU)") + ylim(0.20,0.252) + xlim(0,100) + theme_classic()
ga2 = ggplot(dfr, aes(x = div, y = meangc, col = factor(ngc))) + geom_line() + xlab("Divisions") + ylab("Mean GC events\nper oDNA") + labs(col = "GC rate\n(AU)") + xlim(0,100) + theme_classic()
sf = 3
png("fig-s4-sim-seg.png", width=400*sf, height=150*sf, res=72*sf)
grid.arrange(ga1, ga2, nrow=1)
dev.off()

# second plot -- predicted heteroplasmy distributions for divisions and gene conversion rates
sub = df[(df$div == 1 | df$div == 10 | df$div == 20 | df$div == 50 ),] 
png("fig-s5-sim-pred.png", width=300*sf, height=250*sf, res=72*sf)
ggplot(sub[sub$ngc %in% c(0, 400),], aes(x=h)) + geom_histogram(binwidth=0.1) + facet_grid(div~ngc) + theme_classic()
dev.off()

# data frame to illustate simplified conversion factor between effective population sizes
x = 1:200
df = data.frame(new.n = x, old.n = 50, scale.factor = log(1-1/50)/log(1-1/x))
df = rbind(df, data.frame(new.n = x, old.n = 7, scale.factor = log(1-1/7)/log(1-1/x)))

df$ratio = df$scale.factor / (df$new.n/df$old.n)

png("fig-s6-scalepopsize.png", width=200*sf, height=200*sf, res=72*sf)
ggplot(df, aes(x=new.n, y=scale.factor, col=factor(old.n))) + geom_line() + xlab("New Ne'") + ylab("Scale") + labs(col="Old Ne") + theme_light()
dev.off()


